File: haskell/src/Algorithms/MergeSort.hs
module Algorithms.MergeSort
  ( mergeSort
  , mergeSortList
  ) where

import Data.Vector (Vector)
import qualified Data.Vector as V

-- | Idiomatic Haskell merge sort using Data.Vector (boxed).
-- Matches Rust's Vec<i64> as a contiguous array type.
mergeSort :: Vector Int -> Vector Int
mergeSort xs
  | V.length xs <= 1 = xs
  | otherwise =
      let mid = V.length xs `div` 2
          (left, right) = V.splitAt mid xs
       in merge (mergeSort left) (mergeSort right)

-- | Merge two sorted vectors into one.
merge :: Vector Int -> Vector Int -> Vector Int
merge left right = V.fromList $ go (V.toList left) (V.toList right)
  where
    go [] ys = ys
    go xs [] = xs
    go (x:xs) (y:ys)
      | x <= y    = x : go xs (y:ys)
      | otherwise  = y : go (x:xs) ys

-- | List-based merge sort for comparison.
mergeSortList :: [Int] -> [Int]
mergeSortList [] = []
mergeSortList [x] = [x]
mergeSortList xs =
    let mid = length xs `div` 2
        (left, right) = splitAt mid xs
     in mergeList (mergeSortList left) (mergeSortList right)

mergeList :: [Int] -> [Int] -> [Int]
mergeList [] ys = ys
mergeList xs [] = xs
mergeList (x:xs) (y:ys)
  | x <= y    = x : mergeList xs (y:ys)
  | otherwise  = y : mergeList (x:xs) ys

File: rust/src/algorithms/merge_sort.rs
/// Idiomatic Rust merge sort using Vec and iterators.
/// Pure functional style: creates new Vecs at each level.
pub fn merge_sort(xs: &[i64]) -> Vec<i64> {
    if xs.len() <= 1 {
        return xs.to_vec();
    }
    let mid = xs.len() / 2;
    let left = merge_sort(&xs[..mid]);
    let right = merge_sort(&xs[mid..]);
    merge(&left, &right)
}

fn merge(left: &[i64], right: &[i64]) -> Vec<i64> {
    let mut result = Vec::with_capacity(left.len() + right.len());
    let (mut i, mut j) = (0, 0);

    while i < left.len() && j < right.len() {
        if left[i] <= right[j] {
            result.push(left[i]);
            i += 1;
        } else {
            result.push(right[j]);
            j += 1;
        }
    }
    result.extend_from_slice(&left[i..]);
    result.extend_from_slice(&right[j..]);
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn should_sort_empty_slice() {
        let empty: &[i64] = &[];
        assert_eq!(merge_sort(empty), Vec::<i64>::new());
    }

    #[test]
    fn should_sort_single_element() {
        assert_eq!(merge_sort(&[42]), vec![42]);
    }

    #[test]
    fn should_sort_already_sorted() {
        assert_eq!(merge_sort(&[1, 2, 3, 4, 5]), vec![1, 2, 3, 4, 5]);
    }

    #[test]
    fn should_sort_reversed() {
        assert_eq!(merge_sort(&[5, 4, 3, 2, 1]), vec![1, 2, 3, 4, 5]);
    }

    #[test]
    fn should_sort_with_duplicates() {
        assert_eq!(merge_sort(&[3, 1, 4, 1, 5, 9, 2, 6]), vec![1, 1, 2, 3, 4, 5, 6, 9]);
    }
}
